src/tools/rbd/action/Trash.cc

```diff
@@ -417,53 +417,103 @@ int execute_purge (const po::variables_map &vm,
-      if(arr[i].get_obj()["name"] == pool_name) {
-        json_spirit::mObject stats =  arr[i].get_obj()["stats"].get_obj();
-        pool_percent_used = stats["percent_used"].get_real();
-        if(pool_percent_used <= threshold) {
-          std::cout << "rbd: pool usage is lower than or equal to "
-                    << (threshold*100)
-                    << "%" << endl;
-          std::cout << "Nothing to do" << std::endl;
-          return 0;
-        }
-            
-        pool_total_bytes = stats["max_avail"].get_uint64() +
-                           stats["bytes_used"].get_uint64();
-        break;
-      }
-    }
-    
+
+    std::map<std::string, std::vector<std::string>> datapools;
+
     std::sort(trash_entries.begin(), trash_entries.end(),
-      [](librbd::trash_image_info_t a, librbd::trash_image_info_t b) { 
+      [](librbd::trash_image_info_t a, librbd::trash_image_info_t b) {
         return a.deferment_end_time < b.deferment_end_time;
       }
     );
 
-    uint64_t bytes_to_free = 0;
-    auto bytes_threshold = (uint64_t)(pool_total_bytes * 
-                           (pool_percent_used - threshold));
-    
-    librbd::Image curr_img;
+    std::string data_pool;
     for (const auto& entry : trash_entries) {
-      r = utils::open_image_by_id(io_ctx, entry.id, true, &curr_img);
+      librbd::Image image;
+      r = utils::open_image_by_id(io_ctx, entry.id, true, &image);
       if(r < 0) continue;
+
+      int64_t data_pool_id = image.get_data_pool_id();
+      if (data_pool_id != io_ctx.get_id()) {
+        librados::Rados rados(io_ctx);
+        librados::IoCtx data_io_ctx;
+        r = rados.ioctx_create2(data_pool_id, data_io_ctx);
+        if (r < 0) {
+          data_pool = "<missing data pool " + stringify(data_pool_id) + ">";
+        } else {
+          data_pool = data_io_ctx.get_pool_name();
+          datapools[data_pool];
+          datapools[data_pool].push_back(entry.id);
+        }
+      }
+    }
+    datapools[pool_name]; 
+    uint64_t bytes_to_free;
+
+    for(uint8_t i = 0; i < arr.size(); ++i) {
+      json_spirit::mObject obj = arr[i].get_obj();
+      std::string name = obj.find("name")->second.get_str();
+      std::map<std::string, std::vector<std::string>>::iterator img = datapools.find(name);
+      if(img != datapools.end()) {
+        json_spirit::mObject stats =  arr[i].get_obj()["stats"].get_obj();
+        pool_percent_used = stats["percent_used"].get_real();
+        bytes_to_free = 0;
+        if(pool_percent_used <= threshold) continue;
+            
+        pool_total_bytes = stats["max_avail"].get_uint64() +
+                           stats["bytes_used"].get_uint64();
+          
+        auto bytes_threshold = (uint64_t)(pool_total_bytes * 
+                               (pool_percent_used - threshold));
+    
+        if(arr[i].get_obj()["name"] == pool_name){
+          librbd::Image curr_img;
+          for (const auto& entry : trash_entries) {
```

Comment:
Seems to be duplicating the code below. If you added the trash image ids for the base pool to ```datapools[pool_name]``` above, you could then just iterate through the image ids in the ```datapools[arr[i].get_obj()["name"]]``` to compute which images to remove.